### PR TITLE
Refactor and cleanup footpath-related code

### PR DESCRIPTION
**Pull request by Broxzier**
Mostly replacing harcoded values with defines, renaming some defines to make their purpose clear, and replacing wrongly used defines (and union names) with the correct ones. I've also added a few comments, a few new helper functions (for setting path type for example), and cleaned up some long lines of code.

The only change I made is that the array `DefaultPathSlope` has been cut in half. The last half of its entries were all set to invalid, which makes sense, since for those the diagonal surface flag will be set. Those could be removed, because paths also cannot be placed directly on surfaces with 3 corners raised, which is required for the diagonal flag to be set. (any surface elements that don't follow this format for whatever reason won't cause an issue, since they look funky already anyway).

### CODE REVIEW

File: src/openrct2/interface/viewport_interaction.c [hunk 1]
```diff
@@ -316,7 +316,7 @@ sint32 viewport_interaction_get_item_right(sint32 x, sint32 y, viewport_interact
     case VIEWPORT_INTERACTION_ITEM_FOOTPATH:
         set_map_tooltip_format_arg(0, rct_string_id, STR_MAP_TOOLTIP_STRINGID_CLICK_TO_REMOVE);
         set_map_tooltip_format_arg(2, rct_string_id, STR_FOOTPATH_MAP_TIP);
-        if (tileElement->type & 1)
+        if (tileElement->type & TILE_ELEMENT_PATH_QUEUE_FLAG)
```
Comment (Gymnasiast) [inline src/openrct2/interface/viewport_interaction.c]: There is a call for this.

**Comment by Gymnasiast** [comment]
Generally, I'm happy with refactor PRs. A general thing though: if you're doing tile elements, usually a call is better, in order to reduce dependency on SV6 implementation details (which are most prevalent in tile elements).

**Comment by Broxzier** [comment]
That's what this PR does. Obviously I've missed a few, but using defines instead of magic numbers is still an improvement. Also I didn't know you were refactoring those files too, so there's a lot of conflicts to solve now. 😦

**Comment by Gymnasiast** [comment]
My PR was already there for a few days.

**Comment by Broxzier** [comment]
I know, around the time when I started this. It's just a bummer that I didn't know until your comment. I've rebased now, and made my changes as much in line with yours as I could get them. I'll look over them again tomorrow to finish this PR.

**Comment by Gymnasiast** [comment]
Travis fails because you left one reference to TILE_ELEMENT_WATER_HEIGHT_MASK unchanged - namely in `test/testpaint/compat.c`.

File: src/openrct2-ui/windows/TopToolbar.cpp [hunk 2]
```diff
@@ -1303,7 +1303,7 @@ static void sub_6E1F34(sint16 x, sint16 y, uint16 selected_scenery, sint16* grid
             }
 
             gSceneryPlaceZ = 0;
-            uint16 water_height = tile_element->properties.surface.terrain & TILE_ELEMENT_WATER_HEIGHT_MASK;
+            uint16 water_height = tile_element->properties.surface.terrain & TILE_ELEMENT_SURFACE_WATER_HEIGHT_MASK;
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/TopToolbar.cpp]: Could you create a function `surface_element_get_water_height(const rct_tile_element * surfaceElement)` (or a similar name)?

File: src/openrct2/paint/tile_element/path.c [hunk 3]
```diff
@@ -378,16 +378,18 @@ static void sub_6A4101(paint_session * session, rct_tile_element * tile_element,
...
+        // Draw pole in the front and banner
         boundBoxOffsets.x = BannerBoundBoxes[direction][1].x;
         boundBoxOffsets.y = BannerBoundBoxes[direction][1].y;
         imageId++;
         sub_98197C(session, imageId, 0, 0, 1, 1, 21, height, boundBoxOffsets.x, boundBoxOffsets.y, boundBoxOffsets.z, get_current_rotation());
 
         direction--;
         // If text shown
-        if (direction < 2 && tile_element->properties.path.ride_index != 255 && imageFlags == 0) {
+        if (direction < 2 && tile_element->properties.path.ride_index != RIDE_ENTRY_INDEX_NULL && imageFlags == 0) {
```
Comment (Gymnasiast) [inline src/openrct2/paint/tile_element/path.c]: `RIDE_ID_NULL`. `RIDE_ENTRY_INDEX_NULL` is for `rct_ride_entry`.

File: src/openrct2/world/footpath.c [hunk 4]
```diff
@@ -1773,15 +1786,14 @@ sint32 footpath_is_connected_to_map_edge(sint32 x, sint32 y, sint32 z, sint32 di
 
 bool footpath_element_is_sloped(const rct_tile_element * tileElement)
 {
-    return tileElement->properties.path.type & FOOTPATH_PROPERTIES_FLAG_IS_SLOPED;
+    return (tileElement->properties.path.type & FOOTPATH_PROPERTIES_FLAG_IS_SLOPED) != 0;
 }
 
 void footpath_element_set_sloped(rct_tile_element * tileElement, bool isSloped)
 {
+    tileElement->properties.path.type &= ~FOOTPATH_PROPERTIES_FLAG_IS_SLOPED;
```
Comment (Gymnasiast) [inline src/openrct2/world/footpath.c]: Why are you doing this? It's an unnecessary operation if the footpath is sloped.

File: src/openrct2/world/footpath.c [hunk 5]
```diff
@@ -1791,35 +1803,40 @@ uint8 footpath_element_get_slope_direction(const rct_tile_element * tileElement)
...
     if (isWide)
         tileElement->type |= FOOTPATH_ELEMENT_TYPE_FLAG_IS_WIDE;
-    else
-        tileElement->type &= ~FOOTPATH_ELEMENT_TYPE_FLAG_IS_WIDE;
 }
 
 bool footpath_element_has_path_scenery(const rct_tile_element * tileElement)
 {
-    return (tileElement->properties.path.additions & 0xF) > 0;
+    return footpath_element_get_path_scenery(tileElement) > 0;
```
Comment (Gymnasiast) [inline src/openrct2/world/footpath.c]: I think `!=` would be more consistent with the rest.

File: src/openrct2/world/footpath.h [hunk 6]
```diff
@@ -39,34 +40,55 @@ typedef struct rct_footpath_entry {
 assert_struct_size(rct_footpath_entry, 13);
 #pragma pack(pop)
 
+// Masks for value stored in rct_tile_element.type
```
Comment (Gymnasiast) [inline src/openrct2/world/footpath.h]: value**s**

File: src/openrct2/world/footpath.h [hunk 7]
```diff
@@ -39,34 +40,55 @@ typedef struct rct_footpath_entry {
...
+enum
+{
+    FOOTPATH_ELEMENT_TYPE_FLAG_IS_QUEUE  = (0x1 << 0),
+    FOOTPATH_ELEMENT_TYPE_FLAG_IS_WIDE   = (0x1 << 1),
+    FOOTPATH_ELEMENT_TYPE_DIRECTION_MASK = (0x3 << 6),
```
Comment (Gymnasiast) [inline src/openrct2/world/footpath.h]: I don't find this to be an improvement over my version. I think mine is more readable.

**Comment by Gymnasiast** [comment]
Ok, left my review.

File: src/openrct2/world/footpath.h (hunk 7 above)
Comment (Gymnasiast) [reply src/openrct2/world/footpath.h]: I still find my notation a whole lot easier to read: you know _exactly_ which bits of the values are used.

**Comment by thachachec** [comment]
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!
# LOL PINEAPPLE SEX!

File: src/openrct2/world/footpath.h (hunk 7 above)
Comment (Broxzier) [reply src/openrct2/world/footpath.h]: I've changed it back to your notation now.

**Comment by Gymnasiast** [comment]
Travis fails because TILE_ELEMENT_WATER_HEIGHT_MASK is not defined in testpaint.

**Pull request merged**